gcc/d/dfrontend/expression.c

```diff
@@ -4082,6 +4082,12 @@ Expression *ArrayLiteralExp::semantic(Scope *sc)
         return new ErrorExp();
     }
 
+    if (global.params.noTypeinfo && elements->dim != 0)
+    {
+        error (global.params.noTypeinfo, "Can't use array literal");
```

Comment:
DMD does prevent heap allocation in some places, but not in all it could.  It should at least be consistent with where this error is generated and where `@nogc` generates an error.